code/socialDistribution/views.py

```diff
@@ -384,21 +392,35 @@ def editPost(request, id):
                 post.unlisted = form.cleaned_data.get('unlisted')
                 post.content_media = content_media
 
-                categories = form.cleaned_data.get('categories').split()
-                previousCategories = Category.objects.filter(post=post)
-                previousCategoriesNames = [
-                    cat.category for cat in previousCategories]
-
-                # Create new categories
-                for category in categories:
-                    if category in previousCategoriesNames:
-                        previousCategoriesNames.remove(category)
-                    else:
-                        Category.objects.create(category=category, post=post)
-
-                # Remove old categories that were deleted
-                for category in previousCategoriesNames:
-                    Category.objects.get(category=category, post=post).delete()
+                categories = form.cleaned_data.get('categories')
+
+                if categories is not None:
+                    categories = categories.split()
+                    categories_to_remove = [ cat.category for cat in post.category_set.all()]
+
+                    for category in categories:
+                        try:
+                            existing_category = Category.objects.get(category__iexact=category)
+                            cat_has_post = existing_category.posts.all().filter(id=post.id)
+                        except Category.DoesNotExist:
+                            existing_category = False
+                        
+                        if not existing_category:
+                            post.category_set.create(category=category)
+
+                        elif not cat_has_post:
+                            existing_category.posts.add(post)
+                        
+                        else:
+                            categories_to_remove.remove(existing_category.category)
+
+                    for category in categories_to_remove:
```

Comment:
some line comments would be helpful for the future if we come back and review this code block.